File: python/fpml/core/extract.py
import re
from typing import Any, Optional, cast

from fhirpathpy import evaluate  # type: ignore

from fpml.core.guarded_resource import guarded_resource

from .constants import root_node_key, undefined
from .core_exceptions import FPMLValidationError
from .core_types import (
    Context,
    DictNode,
    FPOptions,
    Matcher,
    MatcherResult,
    Node,
    Path,
    Resource,
    StrNode,
    Transformer,
)
from .utils import flatten, omit_key


def resolve_template(
    resource: Resource,
    template: Any,
    context: Optional[Context] = None,
    fp_options: Optional[FPOptions] = None,
    strict: bool = False,
) -> Any:
    """
    Processes a given template with the specified resource and optional context.

    This function takes a FHIR resource and applies the provided template, optionally
    using a context and additional processing options. The result is a transformed
    representation of the template.

    Args:
        resource (Resource): The input FHIR resource to process.
        template (Any): The template describing the transformation.
        context (Optional[Context], optional): Additional context data. Defaults to None.
        fp_options (Optional[FPOptions], optional): Options for controlling FHIRPath evaluation. Defaults to None.
        strict (bool, optional): Whether to enforce strict mode. Defaults to False.
            See more details on
            [strict mode](https://github.com/beda-software/FHIRPathMappingLanguage/tree/main?tab=readme-ov-file#strict-mode).

    Returns:
        Any: The processed output based on the template.

    Raises:
        FPMLValidationError: If validation of the template or resource fails.

    See Also:
        FHIRPathMappingLanguage Specification:
        https://github.com/beda-software/FHIRPathMappingLanguage/tree/main?tab=readme-ov-file#specification
    """  # noqa: E501
    result = resolve_template_recur(
        [],
        guarded_resource if strict else resource,
        template,
        # Pass resource as context because original is overriden by strict mode
        {"context": resource, **(context or {})},
        fp_options=fp_options,
    )

    return None if result == undefined else result


def resolve_template_recur(
    start_path: Path,
    resource: Resource,
    template: Any,
    context: Context,
    fp_options: Optional[FPOptions] = None,
) -> Any:
    result = iterate_node(
        start_path,
        {root_node_key: template},
        context or {},
        lambda path, node, context: process_node(path, resource, node, context, fp_options),
    )
    if isinstance(result, dict):
        return result.get(root_node_key, undefined)

    return undefined


def process_node(
    path: Path,
    resource: Resource,
    node: Node,
    context: Context,
    fp_options: Optional[FPOptions],
) -> tuple[Node, Context]:
    if isinstance(node, dict):
        new_node, new_context = process_assign_block(path, resource, node, context, fp_options)

        matchers: list[Matcher] = [
            process_context_block,
            process_merge_block,
            process_for_block,
            process_if_block,
        ]

        for matcher in matchers:
            result = matcher(path, resource, new_node, new_context, fp_options)
            if result:
                return result["node"], new_context

        return new_node, new_context

    if isinstance(node, str):
        return process_template_string(path, resource, node, context, fp_options), context

    return node, context


def iterate_node(start_path: Path, node: Node, context: Context, transform: Transformer) -> Node:
    if isinstance(node, list):
        # Arrays are flattened and undefined values are removed here
        cleaned_array = flatten(
            [
                value
                for value in [
                    iterate_node(
                        [*start_path, index],
                        *transform([*start_path, index], value, context),
                        transform,
                    )
                    for index, value in enumerate(node)
                ]
                if value is not undefined
            ]
        )

        return cleaned_array or undefined
    if isinstance(node, dict):
        # undefined values are removed from dicts, but nulls are preserved
        cleaned_object = {
            key: value
            for key, value in {
                key: iterate_node(
                    [*start_path, key],
                    *transform([*start_path, key], value, context),
                    transform,
                )
                for key, value in node.items()
            }.items()
            if value is not undefined
        }

        return cleaned_object or undefined

    return transform(start_path, node, context)[0]


def process_template_string(
    path: Path,
    resource: Resource,
    node: StrNode,
    context: Context,
    fp_options: Optional[FPOptions],
) -> Any:
    array_template_regexp = re.compile(r"{\[\s*([\s\S]+?)\s*\]}")

    match = array_template_regexp.match(node)
    if match:
        expr = match.group(1)
        return evaluate_expression(path, resource, expr, context, fp_options)

    single_template_regexp = re.compile(r"{{\+?\s*([\s\S]+?)\s*\+?}}")
    result = node

    for match in single_template_regexp.finditer(node):
        expr = match.group(1)
        try:
            replacement = evaluate_expression(path, resource, expr, context, fp_options)[0]
        except IndexError:
            return None if match.group(0).startswith("{{+") else undefined
        if match.group(0) == node:
            return replacement
        result = result.replace(match.group(0), str(replacement))

    return result


def process_context_block(
    path: Path,
    resource: Resource,
    node: DictNode,
    context: Context,
    fp_options: Optional[FPOptions],
) -> Optional[MatcherResult]:
    keys = list(node.keys())
    context_regexp = re.compile(r"{{\s*(.+?)\s*}}")
    context_key = next((k for k in keys if context_regexp.match(k)), None)

    if context_key:
        matches = context_regexp.match(context_key)
        expr = matches.group(1) if matches else ""

        if len(keys) > 1:
            raise FPMLValidationError("Context block must be presented as single key", path)

        answers = evaluate_expression(path, resource, expr, context, fp_options)
        return {
            "node": [
                resolve_template_recur(path, answer, node[context_key], context, fp_options)
                for answer in answers
            ]
        }

    return None


def process_for_block(
    path: Path,
    resource: Resource,
    node: DictNode,
    context: Context,
    fp_options: Optional[FPOptions],
) -> Optional[MatcherResult]:
    keys = list(node.keys())

    for_regexp = re.compile(r"{%\s*for\s+(?:(\w+?)\s*,\s*)?(\w+?)\s+in\s+(.+?)\s*%}")
    for_key = next((k for k in keys if for_regexp.match(k)), None)

    if for_key:
        matches = for_regexp.match(for_key)
        if not matches:
            return None

        has_index_key = len(matches.groups()) == 3  # noqa: PLR2004
        index_key = cast(str, matches.group(1)) if has_index_key else None
        item_key = cast(str, matches.group(2) if has_index_key else matches.group(1))
        expr = matches.group(3) if has_index_key else matches.group(2)

        if len(keys) > 1:
            raise FPMLValidationError("For block must be presented as single key", path)

        answers = evaluate_expression(path, resource, expr, context, fp_options)

        return {
            "node": [
                resolve_template_recur(
                    path,
                    resource,
                    node[for_key],
                    {
                        **context,
                        item_key: answer,
                        **({index_key: index} if index_key else {}),
                    },
                    fp_options,
                )
                for index, answer in enumerate(answers)
            ]
        }

    return None


def process_if_block(
    path: Path,
    resource: Resource,
    node: dict[str, Any],
    context: Context,
    fp_options: Optional[FPOptions],
) -> Optional[MatcherResult]:
    keys = list(node.keys())

    if_regexp = re.compile(r"{%\s*if\s+(.+?)\s*%}")
    else_regexp = re.compile(r"{%\s*else\s*%}")

    if_keys = [k for k in keys if if_regexp.match(k)]
    if len(if_keys) > 1:
        raise FPMLValidationError("If block must be presented once", path)
    if_key = if_keys[0] if if_keys else None

    else_keys = [k for k in keys if else_regexp.match(k)]
    if len(else_keys) > 1:
        raise FPMLValidationError("Else block must be presented once", path)
    else_key = else_keys[0] if else_keys else None

    if else_key and not if_key:
        raise FPMLValidationError(
            "Else block must be presented only when if block is presented", path
        )

    if not if_key:
        return None

    matches = if_regexp.match(if_key)
    expr = matches.group(1) if matches else ""

    answer = evaluate_expression(path, resource, f"iif({expr}, true, false)", context, fp_options)[
        0
    ]

    new_node = (
        resolve_template_recur(path, resource, node[if_key], context, fp_options)
        if answer
        else (
            resolve_template_recur(path, resource, node[else_key], context, fp_options)
            if else_key
            else undefined
        )
    )

    is_merge_behavior = len(keys) != (2 if else_key else 1)
    if is_merge_behavior:
        if not isinstance(new_node, dict) and new_node is not None and new_node is not undefined:
            raise FPMLValidationError(
                "If/else block must return object for implicit merge into existing node",
                path,
            )

        return {
            "node": {
                **omit_key(omit_key(node, if_key), else_key),
                **(new_node if isinstance(new_node, dict) else {}),
            }
        }

    return {"node": new_node}


def process_merge_block(
    path: Path,
    resource: Resource,
    node: DictNode,
    context: Context,
    fp_options: Optional[FPOptions],
) -> Optional[MatcherResult]:
    merge_key = next((k for k in node if re.match(r"{%\s*merge\s*%}", k)), None)
    if merge_key:
        merged_node = omit_key(node, merge_key)
        values = node[merge_key] if isinstance(node[merge_key], list) else [node[merge_key]]
        for value in values:
            result = resolve_template_recur(path, resource, value, context, fp_options)
            if not isinstance(result, dict) and result is not None and result is not undefined:
                raise FPMLValidationError("Merge block must contain object", path)
            if result is not undefined and result is not None:
                merged_node.update(result)
        return {"node": merged_node}
    return None


def process_assign_block(
    path: Path,
    resource: Resource,
    node: DictNode,
    context: Context,
    fp_options: Optional[FPOptions],
) -> tuple[DictNode, Context]:
    extended_context = context.copy()
    assign_key = next((k for k in node if re.match(r"{%\s*assign\s*%}", k)), None)
    if assign_key:
        # TODO: re-write without copy-pasting
        if isinstance(node[assign_key], list):
            for obj in node[assign_key]:
                if len(obj) != 1:
                    raise FPMLValidationError(
                        "Assign block must accept only one key per object", path
                    )
                result = {
                    key: resolve_template_recur(
                        [*path, key], resource, obj_value, extended_context, fp_options
                    )
                    for key, obj_value in obj.items()
                }
                key = next(iter(obj.keys()))
                extended_context.update({key: result[key] if result[key] != undefined else None})
        elif isinstance(node[assign_key], dict) and len(node[assign_key]) == 1:
            obj = node[assign_key]
            result = {
                key: resolve_template_recur(
                    [*path, key], resource, obj_value, extended_context, fp_options
                )
                for key, obj_value in obj.items()
            }
            key = next(iter(obj.keys()))
            extended_context.update({key: result[key] if result[key] != undefined else None})
        else:
            raise FPMLValidationError("Assign block must accept array or object", path)
        return omit_key(node, assign_key), extended_context
    return node, context


def evaluate_expression(
    path: Path,
    resource: Resource,
    expression: str,
    context: Context,
    fp_options: Optional[FPOptions] = None,
) -> list[Any]:
    fp_options_copy = cast(dict, fp_options or {}).copy()
    model = fp_options_copy.pop("model", None)

    try:
        return evaluate(resource, expression, context, model, options=fp_options_copy)
    except Exception as exc:
        raise FPMLValidationError(f"Cannot evaluate '{expression}': {exc}", path) from exc

File: ts/server/src/core/extract.ts
import * as fhirpath from 'fhirpath';

type Resource = Record<string, any>;
type Path = Array<string | number>;

// TODO: looks a bit hacky to use extra node here
// TODO: I believe it might be re-written without using it
const rootNodeKey = '__rootNode__';

export interface FPOptions {
    userInvocationTable?: UserInvocationTable;
}

export class FPMLValidationError extends Error {
    errorPath: string;
    errorMessage: string;

    constructor(message: string, path: Path) {
        const pathStr = path.filter((x) => x != rootNodeKey).join('.');
        super(`${message}. Path '${pathStr}'`);
        this.errorMessage = message;
        this.errorPath = pathStr;
    }
}

const guardedResourceFactory = (resource: any) =>
    new Proxy(resource, {
        get: (obj, prop) => {
            if (prop === '__path__') {
                return undefined;
            }

            if (prop === 'resourceType') {
                return obj.resourceType;
            }

            throw new Error(
                `Forbidden access to resource property ${String(
                    prop,
                )} in strict mode. Use context instead`,
            );
        },
    });

export function resolveTemplate(
    resource: Resource,
    template: any,
    context?: Context,
    model?: Model,
    fpOptions?: FPOptions,
    strict?: boolean,
): any {
    return resolveTemplateRecur(
        [],
        strict ? guardedResourceFactory(resource) : resource,
        template,
        { context: resource, ...(context ?? {}) },
        model,
        fpOptions,
    );
}

function resolveTemplateRecur(
    startPath: Path,
    resource: Resource,
    template: any,
    initialContext: Context,
    model?: Model,
    fpOptions?: FPOptions,
): any {
    return iterateObject(
        startPath,
        { [rootNodeKey]: template },
        initialContext,
        (path, node, context) => {
            if (isPlainObject(node)) {
                const { node: newNode, context: newContext } = processAssignBlock(
                    path,
                    resource,
                    node,
                    context,
                    model,
                    fpOptions,
                );
                const matchers = [
                    processContextBlock,
                    processMergeBlock,
                    processForBlock,
                    processIfBlock,
                ];
                for (const matcher of matchers) {
                    const result = matcher(path, resource, newNode, newContext, model, fpOptions);

                    if (result) {
                        return { node: result.node, context: newContext };
                    }
                }

                return { node: newNode, context: newContext };
            } else if (typeof node === 'string') {
                return {
                    node: processTemplateString(path, resource, node, context, model, fpOptions),
                    context,
                };
            }

            return { node, context };
        },
    )?.[rootNodeKey];
}

function processTemplateString(
    path: Path,
    resource: Resource,
    node: string,
    context: Context,
    model: Model,
    fpOptions: FPOptions,
) {
    let match:
        | RegExpExecArray
        | { [Symbol.replace](string: string, replaceValue: string): string }[];

    const arrayTemplateRegExp = /{\[\s*([\s\S]+?)\s*\]}/g;
    match = arrayTemplateRegExp.exec(node);
    if (match) {
        const expr = match[1];

        return evaluateExpression(path, resource, expr, context, model, fpOptions);
    }

    const singleTemplateRegExp = /{{\+?\s*([\s\S]+?)\s*\+?}}/g;
    let result: any = node;

    while ((match = singleTemplateRegExp.exec(node)) !== null) {
        const expr = match[1];
        const replacement = evaluateExpression(path, resource, expr, context, model, fpOptions)[0];

        if (replacement === undefined) {
            if (match[0].startsWith('{{+')) {
                return null;
            }

            return undefined;
        }

        if (match[0] === node) {
            return replacement;
        }

        result = result.replace(match[0], replacement);
    }

    return result;
}

function processAssignBlock(
    path: Path,
    resource: Resource,
    node: any,
    context: Context,
    model: Model,
    fpOptions: FPOptions,
): { node: any; context: Context } {
    const extendedContext = { ...context };
    const keys = Object.keys(node);
    const assignRegExp = /{%\s*assign\s*%}/;
    const assignKey = keys.find((k) => k.match(assignRegExp));
    if (assignKey) {
        if (Array.isArray(node[assignKey])) {
            node[assignKey].forEach((obj) => {
                if (Object.keys(obj).length !== 1) {
                    throw new FPMLValidationError(
                        'Assign block must accept only one key per object',
                        path,
                    );
                }

                return Object.entries(
                    mapValues(obj, (objValue, objKey) =>
                        resolveTemplateRecur(
                            [...path, objKey],
                            resource,
                            objValue,
                            extendedContext,
                            model,
                            fpOptions,
                        ),
                    ),
                ).forEach(([key, value]) => {
                    extendedContext[key] = value;
                });
            });
        } else if (isPlainObject(node[assignKey])) {
            if (Object.keys(node[assignKey]).length !== 1) {
                throw new FPMLValidationError(
                    'Assign block must accept only one key per object',
                    path,
                );
            }
            Object.entries(
                mapValues(node[assignKey], (objValue, objKey) =>
                    resolveTemplateRecur(
                        [...path, objKey],
                        resource,
                        objValue,
                        extendedContext,
                        model,
                        fpOptions,
                    ),
                ),
            ).forEach(([key, value]) => {
                extendedContext[key] = value;
            });
        } else {
            throw new FPMLValidationError('Assign block must accept array or object', path);
        }

        return { node: omitKey(node, assignKey), context: extendedContext };
    }

    return { node, context };
}

function processMergeBlock(
    path: Path,
    resource: Resource,
    node: any,
    context: Context,
    model: Model,
    fpOptions: FPOptions,
): { node: any } | undefined {
    const keys = Object.keys(node);

    const mergeRegExp = /{%\s*merge\s*%}/;
    const mergeKey = keys.find((k) => k.match(mergeRegExp));

    if (mergeKey) {
        return {
            node: (Array.isArray(node[mergeKey]) ? node[mergeKey] : [node[mergeKey]]).reduce(
                (mergeAcc, nodeValue) => {
                    const result = resolveTemplateRecur(
                        path,
                        resource,
                        nodeValue,
                        context,
                        model,
                        fpOptions,
                    );
                    if (!isPlainObject(result) && result !== null && result !== undefined) {
                        throw new FPMLValidationError('Merge block must contain object', path);
                    }

                    return { ...mergeAcc, ...(isPlainObject(result) ? result : {}) };
                },
                omitKey(node, mergeKey),
            ),
        };
    }
}

function processForBlock(
    path: Path,
    resource: Resource,
    node: any,
    context: Context,
    model: Model,
    fpOptions: FPOptions,
): { node: any } | undefined {
    const keys = Object.keys(node);

    const forRegExp = /{%\s*for\s+(?:(\w+?)\s*,\s*)?(\w+?)\s+in\s+(.+?)\s*%}/;
    const forKey = keys.find((k) => k.match(forRegExp));

    if (forKey) {
        const matches = forKey.match(forRegExp);
        const hasIndexKey = matches.length === 4;
        const indexKey = hasIndexKey ? matches[1] : null;
        const itemKey = hasIndexKey ? matches[2] : matches[1];
        const expr = hasIndexKey ? matches[3] : matches[2];

        if (keys.length > 1) {
            throw new FPMLValidationError(`For block must be presented as single key`, path);
        }

        const answers = evaluateExpression(path, resource, expr, context, model, fpOptions);
        return {
            node: answers.map((answer, index) =>
                resolveTemplateRecur(
                    path,
                    resource,
                    node[forKey],
                    {
                        ...context,
                        [itemKey]: answer,
                        ...(hasIndexKey ? { [indexKey]: index } : {}),
                    },
                    model,
                    fpOptions,
                ),
            ),
        };
    }
}

function processContextBlock(
    path: Path,
    resource: Resource,
    node: any,
    context: Context,
    model: Model,
    fpOptions: FPOptions,
): { node: any } | undefined {
    const keys = Object.keys(node);

    const contextRegExp = /{{\s*(.+?)\s*}}/;
    const contextKey = keys.find((k) => k.match(contextRegExp));
    if (contextKey) {
        const matches = contextKey.match(contextRegExp);
        const expr = matches[1];

        if (keys.length > 1) {
            throw new FPMLValidationError('Context block must be presented as single key', path);
        }

        const answers = evaluateExpression(path, resource, expr, context, model, fpOptions);
        const result: any[] = answers.map((answer) =>
            resolveTemplateRecur(path, answer, node[contextKey], context, model, fpOptions),
        );

        return { node: result };
    }
}

function processIfBlock(
    path: Path,
    resource: Resource,
    node: any,
    context: Context,
    model: Model,
    fpOptions: FPOptions,
): { node: any } | undefined {
    const keys = Object.keys(node);

    const ifRegExp = /{%\s*if\s+(.+?)\s*%}/;
    const elseRegExp = /{%\s*else\s*%}/;

    const ifKeys = keys.filter((k) => k.match(ifRegExp));
    if (ifKeys.length > 1) {
        throw new FPMLValidationError('If block must be presented once', path);
    }
    const ifKey = ifKeys[0];

    const elseKeys = keys.filter((k) => k.match(elseRegExp));
    if (elseKeys.length > 1) {
        throw new FPMLValidationError('Else block must be presented once', path);
    }
    const elseKey = elseKeys[0];

    if (elseKey && !ifKey) {
        throw new FPMLValidationError(
            'Else block must be presented only when if block is presented',
            path,
        );
    }

    if (ifKey) {
        const matches = ifKey.match(ifRegExp);
        const expr = matches[1];

        const answer = evaluateExpression(
            path,
            resource,
            `iif(${expr}, true, false)`,
            context,
            model,
            fpOptions,
        )[0];

        const newNode = answer
            ? resolveTemplateRecur(path, resource, node[ifKey], context, model, fpOptions)
            : elseKey
            ? resolveTemplateRecur(path, resource, node[elseKey], context, model, fpOptions)
            : undefined;

        const isMergeBehavior = keys.length !== (elseKey ? 2 : 1);
        if (isMergeBehavior) {
            if (!isPlainObject(newNode) && newNode !== null && newNode !== undefined) {
                throw new FPMLValidationError(
                    'If/else block must return object for implicit merge into existing node',
                    path,
                );
            }

            return {
                node: {
                    ...omitKey(omitKey(node, ifKey), elseKey),
                    ...(isPlainObject(newNode) ? newNode : {}),
                },
            };
        }
        return { node: newNode };
    }
}

type Transformer = (path: Path, node: any, context: Context) => { node: any; context: Context };

function iterateObject(startPath: Path, obj: any, context: Context, transform: Transformer): any {
    if (Array.isArray(obj)) {
        // Arrays are flattened and null/undefined values are removed here
        const cleanedArray = obj
            .flatMap((value, index) => {
                const result = transform([...startPath, index], value, context);

                return iterateObject(
                    [...startPath, index],
                    result.node,
                    result.context,
                    transform,
                );
            })
            .filter((x) => x !== null && x !== undefined);
        return cleanedArray.length ? cleanedArray : undefined;
    } else if (isPlainObject(obj)) {
        const objResult = mapValues(obj, (value, key) => {
            const result = transform([...startPath, key], value, context);

            return iterateObject([...startPath, key], result.node, result.context, transform);
        });

        const cleanedObject = filterValues(objResult, (_key, value) => value !== undefined);

        return Object.entries(cleanedObject).length ? cleanedObject : undefined;
    }

    return transform(startPath, obj, context).node;
}

function isPlainObject(obj: any) {
    return Object.prototype.toString.call(obj) === '[object Object]';
}

function mapValues(obj: object, fn: (value: any, key: string) => any) {
    return Object.fromEntries(
        Object.entries(obj).map(([key, value]) => {
            return [key, fn(value, key)];
        }),
    );
}

function filterValues(obj: object, fn: (value: any, key: string) => any) {
    return Object.fromEntries(
        Object.entries(obj).filter(([key, value]) => {
            return [key, fn(value, key)];
        }),
    );
}

function omitKey(obj: any, key: string) {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { [key]: _, ...rest } = obj;

    return rest;
}

export function evaluateExpression(
    path: Path,
    resource: any,
    expression: string,
    context: Context,
    model: Model,
    options: FPOptions,
) {
    try {
        return fhirpath.evaluate(
            resource,
            expression,
            // fhirpath mutates context https://github.com/HL7/fhirpath.js/issues/155
            { ...context },
            model,
            options,
        );
    } catch (exc) {
        throw new FPMLValidationError(`Can not evaluate '${expression}': ${exc}`, path);
    }
}
